"chrM" not being excluded

Hi, 

I get an error when I run hicrep on my mouse HiC mcool files (mm10 genome). Basically, the lack of data for chrM is causing hicrep to fail. I see that you have rule in the code to exclude chromosome "M", but not "chrM" (line 161 hicrep.py). Would it be possible to allow for more general nomenclature of the mitochondrial chromosome?  

Right now, I get around the issue by explicitly listing all chromosomes as an argument but would rather not need to do that.

**Error is shown below: Python script uses the toy example from github README with my own mcool files** 
```
>python hicrep_test.py 
Traceback (most recent call last):
  File "hicrep_test.py", line 30, in <module>
    scc = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample)
  File "/home/kevbrick/data/get_sequencing_data/hicrep/lib/python3.8/site-packages/hicrep/hicrep.py", line 161, in hicrepSCC
    assert mS1.size > 0, "Contact matrix 1 of chromosome %s is empty" % (chrName)
AssertionError: Contact matrix 1 of chromosome chrM is empty
```

Thanks for the great tool !
Kevin

--- a/hicrep/__init__.py
+++ b/hicrep/__init__.py
@@ -52,6 +52,10 @@ def main(*args):
     parser.add_argument("--chrNames", type=str, nargs='*', default=[],
                         help="Only compute the SCC scores on this subset of\
                         chromosomes whose names are provided")
+    parser.add_argument("--excludeChr", type=str, nargs='*', default=[],
+                        help="Exclude chromosomes from the SCC score calculations.\
+                        Mitochondrial chromosomes named \"M\" are excluded by\
+                        default.")
 
     args = parser.parse_args()
 
@@ -80,10 +84,11 @@ def main(*args):
     dBPMax = args.dBPMax
     bDownSample = args.bDownSample
     chrNames = np.array(args.chrNames, dtype=str)
-
+    excludeChr = np.array(args.excludeChr, dtype=str)
+    
     cool1, binSize1 = readMcool(fmcool1, binSize)
     cool2, binSize2 = readMcool(fmcool2, binSize)
 
-    scc = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNames)
+    scc = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNames, excludeChr)
 
     np.savetxt(fout, scc, "%30.15e", header=header)

--- a/hicrep/__init__.py
+++ b/hicrep/__init__.py
@@ -51,18 +51,22 @@ def main(*args):
                         number of contacts.")
     parser.add_argument("--chrNames", type=str, nargs='*', default=[],
                         help="Only compute the SCC scores on this subset of\
-                        chromosomes whose names are provided")
+                        chromosomes whose names are provided. The output SCC\
+                        scores will be ordered as the input chromosome names\
+                        here")
     parser.add_argument("--excludeChr", type=str, nargs='*', default=['M'],
                         help="Exclude chromosomes from the SCC score calculations.\
                         Mitochondrial chromosomes named \"M\" are excluded by\
-                        default.")
+                        default. The output SCC scores will be ordered as the\
+                        chromosomes in the input Cooler files by removing those\
+                        chromosomes provided here")
 
     args = parser.parse_args()
 
-    if not args.excludeChr == ['M']:
-        assert not (args.chrNames and args.excludeChr), f"""
-            Please use --chrNames OR --excludeChr arguments but not both."""
-    
+    assert not (args.excludeChr != ['M'] and len(args.chrNames) > 0), f"""
+        Please use --chrNames OR --excludeChr arguments but not both.
+        """
+
     header = "#"+" ".join(sys.argv)+"\n"
 
     # Check if current script is under revision control
@@ -87,12 +91,18 @@ def main(*args):
     h = args.h
     dBPMax = args.dBPMax
     bDownSample = args.bDownSample
-    chrNames = np.array(args.chrNames, dtype=str)
-    excludeChr = np.array(args.excludeChr, dtype=str)
-    
+    chrNames = args.chrNames
+    excludeChr = set(args.excludeChr)
+
+    if len(excludeChr) != len(args.excludeChr):
+        warnings.warn(f"""
+            Duplicate excludeChr found. Please remove them in --excludeChr""")
+
     cool1, binSize1 = readMcool(fmcool1, binSize)
     cool2, binSize2 = readMcool(fmcool2, binSize)
 
-    scc = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNames, excludeChr)
+    scc = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample,
+                    chrNames if len(chrNames) > 0 else None,
+                    excludeChr if len(excludeChr) > 0 else None)
 
     np.savetxt(fout, scc, "%30.15e", header=header)

--- a/hicrep/__init__.py
+++ b/hicrep/__init__.py
@@ -52,13 +52,17 @@ def main(*args):
     parser.add_argument("--chrNames", type=str, nargs='*', default=[],
                         help="Only compute the SCC scores on this subset of\
                         chromosomes whose names are provided")
-    parser.add_argument("--excludeChr", type=str, nargs='*', default=[],
+    parser.add_argument("--excludeChr", type=str, nargs='*', default=['M'],
                         help="Exclude chromosomes from the SCC score calculations.\
                         Mitochondrial chromosomes named \"M\" are excluded by\
                         default.")
 
     args = parser.parse_args()
 
+    if not args.excludeChr == ['M']:
+        assert not (args.chrNames and args.excludeChr), f"""
+            Please use --chrNames OR --excludeChr arguments but not both."""
+    
     header = "#"+" ".join(sys.argv)+"\n"
 
     # Check if current script is under revision control

--- a/hicrep/hicrep.py
+++ b/hicrep/hicrep.py
@@ -93,7 +93,8 @@ def sccByDiag(m1: sp.coo_matrix, m2: sp.coo_matrix, nDiags: int):
 
 def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
               h: int, dBPMax: int, bDownSample: bool,
-              chrNames: np.ndarray = np.array([], dtype=str)):
+              chrNames: np.ndarray = np.array([], dtype=str),
+              excludeChr: np.ndarray = np.array([], dtype=str)):
     """Compute hicrep score between two input Cooler contact matrices
 
     Args:
@@ -108,6 +109,9 @@ def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
         chrNames: `np.ndarray` Numpy array of chromosome names whose SCC to
         compute. Default to empty array, which means all chromosomes in the
         genome are used to compute SCC
+        excludeChr: `np.ndarray` Numpy array of chromosome names to exclude
+        from SCC computation. Default to empty array. Chromosome "M" is
+        always excluded.
 
     Returns:
         `float` scc scores for each chromosome
@@ -151,8 +155,9 @@ def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
     n2 = coolerInfo(cool2, 'sum')
     if chrNames.size == 0:
         chrNames = cool1.chroms()[:]['name'].to_numpy()
-    # filter out mitochondria chromosome
-    chrNames = np.array([name for name in chrNames if name != 'M'])
+    # filter out mitochondria chromosome and other excluded chromosomes
+    excludeChr = np.append(excludeChr,'M')
+    chrNames = np.array([name for name in chrNames if name not in excludeChr])
     scc = np.full(chrNames.shape[0], -2.0)
     for iChr in range(chrNames.shape[0]):
         chrName = chrNames[iChr]

--- a/hicrep/hicrep.py
+++ b/hicrep/hicrep.py
@@ -93,8 +93,7 @@ def sccByDiag(m1: sp.coo_matrix, m2: sp.coo_matrix, nDiags: int):
 
 def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
               h: int, dBPMax: int, bDownSample: bool,
-              chrNames: np.ndarray = np.array([], dtype=str),
-              excludeChr: np.ndarray = np.array([], dtype=str)):
+              chrNames: list = None, excludeChr: set = None):
     """Compute hicrep score between two input Cooler contact matrices
 
     Args:
@@ -106,11 +105,11 @@ def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
         distance (bp) away
         bDownSample: `bool` Down sample the input with more contacts
         to the same number of contacts as in the other input
-        chrNames: `np.ndarray` Numpy array of chromosome names whose SCC to
-        compute. Default to empty array, which means all chromosomes in the
+        chrNames: `list` List of chromosome names whose SCC to
+        compute. Default to None, which means all chromosomes in the
         genome are used to compute SCC
-        excludeChr: `np.ndarray` Numpy array of chromosome names to exclude
-        from SCC computation. Default to empty array.
+        excludeChr: `set` Set of chromosome names to exclude from SCC
+        computation. Default to None.
 
     Returns:
         `float` scc scores for each chromosome
@@ -152,13 +151,23 @@ def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
     # get the total number of contacts as normalizing constant
     n1 = coolerInfo(cool1, 'sum')
     n2 = coolerInfo(cool2, 'sum')
-    if chrNames.size == 0:
-        chrNames = cool1.chroms()[:]['name'].to_numpy()
-    # filter out mitochondria chromosome and other excluded chromosomes
-    chrNames = np.array([name for name in chrNames if name not in excludeChr])
-    scc = np.full(chrNames.shape[0], -2.0)
-    for iChr in range(chrNames.shape[0]):
-        chrName = chrNames[iChr]
+    # Use dict here so that the chrNames don't duplicate
+    if chrNames is None:
+        chrNamesDict = dict.fromkeys(cool1.chroms()[:]['name'].tolist())
+    else:
+        chrNamesDict = dict.fromkeys(chrNames)
+    # It's important to preserve the order of the input chrNames so that the
+    # user knows the order of the output SCC scores so we bail when encounter
+    # duplicate names rather than implicit prunning the names.
+    assert chrNames is None or len(chrNamesDict) == len(chrNames), f"""
+        Found Duplicates in {chrNames}. Please remove them.
+        """
+    # filter out excluded chromosomes
+    if excludeChr is None:
+        excludeChr = set()
+    chrNames = [ chrName for chrName in chrNamesDict if chrName not in excludeChr ]
+    scc = np.full(len(chrNames), -2.0)
+    for iChr, chrName in enumerate(chrNames):
         # normalize by total number of contacts
         mS1 = getSubCoo(p1, bins1, chrName)
         assert mS1.size > 0, "Contact matrix 1 of chromosome %s is empty" % (chrName)

--- a/hicrep/hicrep.py
+++ b/hicrep/hicrep.py
@@ -110,8 +110,7 @@ def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
         compute. Default to empty array, which means all chromosomes in the
         genome are used to compute SCC
         excludeChr: `np.ndarray` Numpy array of chromosome names to exclude
-        from SCC computation. Default to empty array. Chromosome "M" is
-        always excluded.
+        from SCC computation. Default to empty array.
 
     Returns:
         `float` scc scores for each chromosome
@@ -156,7 +155,6 @@ def hicrepSCC(cool1: cooler.api.Cooler, cool2: cooler.api.Cooler,
     if chrNames.size == 0:
         chrNames = cool1.chroms()[:]['name'].to_numpy()
     # filter out mitochondria chromosome and other excluded chromosomes
-    excludeChr = np.append(excludeChr,'M')
     chrNames = np.array([name for name in chrNames if name not in excludeChr])
     scc = np.full(chrNames.shape[0], -2.0)
     for iChr in range(chrNames.shape[0]):

--- a/tests/test_HiCRepSCC.py
+++ b/tests/test_HiCRepSCC.py
@@ -47,26 +47,27 @@ def testHumanHiC():
 
     # Test the computation of a subset of chromosomes give the same results as
     # the whole set
-    chrNames = np.array(['chr1', 'chr2', 'chrX'], dtype=str)
+    chrNames = ['chr1', 'chr2', 'chrX']
     resultsSub = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNames)
-    chrNamesAll = cool1.chroms()[:]['name'].to_numpy()
-    chrNamesAll = np.array([name for name in chrNamesAll if name != 'M'])
+    chrNamesAll = cool1.chroms()[:]['name'].tolist()
+    chrNamesAll = [name for name in chrNamesAll if name != 'M']
     iChrs = np.where(np.isin(chrNamesAll, chrNames))[0]
     assert (results[iChrs] == resultsSub).all(), f"""
         SCC scores between {fmcool1} and {fmcool2} on chromosome subset
         {chrNames} differ from those computed from the whole set. The whole
         genome results are: {results} and the subset indices are {iChrs}.
         """
-        
+
     # Test the computation when excluding chromosomes give the same results as
     # the whole set
-    exclNames = np.array(['chr1', 'chr2', 'chrX'], dtype=str)
-    resultsExcl = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNamesAll, exclNames)
-    chrExclNames = np.array([name for name in chrNamesAll if name not in exclNames])
-    ieChrs = np.where(np.isin(chrNamesAll, chrExclNames))[0]
+    exclNames = set(['chr1', 'chr2', 'chrX'])
+    resultsExcl = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample,
+                            chrNamesAll, exclNames)
+    chrNamesRemain = [name for name in chrNamesAll if name not in exclNames]
+    ieChrs = np.where(np.isin(chrNamesAll, chrNamesRemain))[0]
     assert (results[ieChrs] == resultsExcl).all(), f"""
-        SCC scores between {fmcool1} and {fmcool2} differ from those computed 
-        from the whole set when {chrExclNames} are excluded. The whole genome 
+        SCC scores between {fmcool1} and {fmcool2} differ from those computed
+        from the whole set when {chrNamesRemain} are excluded. The whole genome
         results are: {results} and the subset indices are {iChrs}.
         """
 
@@ -97,25 +98,26 @@ def testFlyHiC():
 
     # Test the computation of a subset of chromosomes give the same results as
     # the whole set
-    chrNames = np.array(['chr2L', 'chr2R', 'chrX'], dtype=str)
+    chrNames = ['chr2L', 'chr2R', 'chrX']
     resultsSub = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNames)
-    chrNamesAll = cool1.chroms()[:]['name'].to_numpy()
-    chrNamesAll = np.array([name for name in chrNamesAll if name != 'M'])
+    chrNamesAll = cool1.chroms()[:]['name'].tolist()
+    chrNamesAll = [name for name in chrNamesAll if name != 'M']
     iChrs = np.where(np.isin(chrNamesAll, chrNames))[0]
     assert (results[iChrs] == resultsSub).all(), f"""
         SCC scores between {fmcool1} and {fmcool2} on chromosome subset
         {chrNames} differ from those computed from the whole set. The whole
         genome results are: {results} and the subset indices are {iChrs}.
         """
-    
+
     # Test the computation when excluding chromosomes give the same results as
     # the whole set
-    exclNames = np.array(['chr2L', 'chr2R', 'chrX'], dtype=str)
-    resultsExcl = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNamesAll, exclNames)
-    chrExclNames = np.array([name for name in chrNamesAll if name not in exclNames])
-    ieChrs = np.where(np.isin(chrNamesAll, chrExclNames))[0]
+    exclNames = set(['chr2L', 'chr2R', 'chrX'])
+    resultsExcl = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample,
+                            chrNamesAll, exclNames)
+    chrNamesRemain = [name for name in chrNamesAll if name not in exclNames]
+    ieChrs = np.where(np.isin(chrNamesAll, chrNamesRemain))[0]
     assert (results[ieChrs] == resultsExcl).all(), f"""
-        SCC scores between {fmcool1} and {fmcool2} differ from those computed 
-        from the whole set when {chrExclNames} are excluded. The whole genome 
+        SCC scores between {fmcool1} and {fmcool2} differ from those computed
+        from the whole set when {chrNamesRemain} are excluded. The whole genome
         results are: {results} and the subset indices are {iChrs}.
         """

--- a/tests/test_HiCRepSCC.py
+++ b/tests/test_HiCRepSCC.py
@@ -57,6 +57,18 @@ def testHumanHiC():
         {chrNames} differ from those computed from the whole set. The whole
         genome results are: {results} and the subset indices are {iChrs}.
         """
+        
+    # Test the computation when excluding chromosomes give the same results as
+    # the whole set
+    exclNames = np.array(['chr1', 'chr2', 'chrX'], dtype=str)
+    resultsExcl = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNamesAll, exclNames)
+    chrExclNames = np.array([name for name in chrNamesAll if name not in exclNames])
+    ieChrs = np.where(np.isin(chrNamesAll, chrExclNames))[0]
+    assert (results[ieChrs] == resultsExcl).all(), f"""
+        SCC scores between {fmcool1} and {fmcool2} differ from those computed 
+        from the whole set when {chrExclNames} are excluded. The whole genome 
+        results are: {results} and the subset indices are {iChrs}.
+        """
 
 
 def testFlyHiC():
@@ -94,4 +106,16 @@ def testFlyHiC():
         SCC scores between {fmcool1} and {fmcool2} on chromosome subset
         {chrNames} differ from those computed from the whole set. The whole
         genome results are: {results} and the subset indices are {iChrs}.
+        """
+    
+    # Test the computation when excluding chromosomes give the same results as
+    # the whole set
+    exclNames = np.array(['chr2L', 'chr2R', 'chrX'], dtype=str)
+    resultsExcl = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample, chrNamesAll, exclNames)
+    chrExclNames = np.array([name for name in chrNamesAll if name not in exclNames])
+    ieChrs = np.where(np.isin(chrNamesAll, chrExclNames))[0]
+    assert (results[ieChrs] == resultsExcl).all(), f"""
+        SCC scores between {fmcool1} and {fmcool2} differ from those computed 
+        from the whole set when {chrExclNames} are excluded. The whole genome 
+        results are: {results} and the subset indices are {iChrs}.
         """

